### PR TITLE
Reduce errors preventing role, field from passing from invitation to user

### PR DESCRIPTION
**Pull request by AddisonDunn**
## Product Description

[Jira](https://dimagi.atlassian.net/browse/USH-4727?atlOrigin=eyJpIjoiODE0ZDcxYWFlODE0NDAzZGJmNDVjNTFkMGE0MzlhZWEiLCJwIjoiaiJ9).

This PR aims to reduce the frequency of a problem where invited web users are created without a role, profile, etc.

## Technical Summary

When a user accepts an invite, an exception is sometimes triggered. In [the past 90 days](https://dimagi.sentry.io/issues/?environment=production&project=136860&query=is%3Aunresolved+issue.priority%3A%5Bhigh%2C+medium%5D+settings%2Fusers%2Fjoin&referrer=issue-list&statsPeriod=90d&utc=true), this has been due to either:

1. A ResourceConflict when saving the user/interacting with Couch
2. An issue communicating with the Tableau server

When an error is triggered in `add_as_web_user`, it seems to stop execution _after_ the `CouchUser` was created but _before_ the fields were populated from the invitation object, causing the end-user issue. So this PR attempts to reduce those errors by 1) reducing the number of saves to couch and 2) passing over any issues communicating with Tableau, since those discrepancies should be cleaned up on the next hourly sync anyways.

There seems to be an existing history of these `ResourceConflict` issues (possibly specific to _users_), and I think is all a bit beyond my knowledge of Couch/couchdbkit. I do think reducing the calls to couch has a good chance of reducing or completely eliminating these errors, though. 

## Feature Flag

None, though some changes are behind TABLEAU_USER_SYNCING

### Safety story

- Tested locally to ensure user and associated Tableau user is created with proper information
- Existing tests cover Tableau user syncing methods well I think, and added new automated tests for user invitation view/workflow
- Should be mostly a refactor. I expect the main risks are with the live prod environment--the `ResourceConflict`s are still partially a mystery to me
- If a `ResourceConflict` is triggered upon initial user save, it will block the user from logging in (because they will have Django user associated with their email but no `CouchUser`). But this should be an existing issue

### Automated test coverage

- Some exist for the Tableau user syncing methods
- Added new test(s) for the invitation workflow

### QA Plan

Not planning to do QA

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/users/tests/test_user_invitation.py [hunk 1]
```diff
@@ -23,30 +26,45 @@ def cleaned_data(self):
 
 class TestUserInvitation(TestCase):
 
+    @classmethod
+    def setUpTestData(cls):
+        # super(TestCase, cls).setUpTestData()
```
Comment (Jtang-1) [inline corehq/apps/users/tests/test_user_invitation.py:35]: nit: can delete this

File: corehq/apps/reports/util.py [hunk 2]
```diff
@@ -512,23 +512,29 @@ def get_matching_tableau_users_from_other_domains(user):
...
+        except TableauConnectedApp.DoesNotExist as e:
+            _notify_tableau_exception(e, domain)
+            return
+        user, created, matching_tableau_users_from_other_domains_exist = _add_tableau_user_local(session, username)
+        if created and not matching_tableau_users_from_other_domains_exist:
+            try:
+                _add_tableau_user_remote(session, user)
+            except TableauAPIError as e:
+                if e.code != 409017:  # This is the "user already added to site" code.
+                    raise
```
Comment (Jtang-1) [inline corehq/apps/reports/util.py:532]: should this also check for blocking exception?

File: corehq/apps/reports/util.py [hunk 3]
```diff
@@ -512,23 +512,29 @@ def get_matching_tableau_users_from_other_domains(user):
 
 
 @atomic
-def add_tableau_user(domain, username):
+def add_tableau_user(domain, username, blocking_exception=False):
```
Comment (Jtang-1) [inline corehq/apps/reports/util.py:515]: I noticed `blocking_exception=True` wasn't passed as arguments to where `add_tableau_user` is called (such as in `add_domain_membership` and `undo`). Couldn't it also be an issue there?

Also, just a thought - would it make sense to handle the exception on the calling function as opposed to introducing this parameter?

File: corehq/apps/users/models.py [hunk 4]
```diff
@@ -562,8 +562,8 @@ def add_as_web_user(self, domain, role, primary_location_id=None,
             self.get_domain_membership(domain).program_id = program_id
         if domain_obj.uses_locations:
             if primary_location_id:
-                self.set_location(domain, primary_location_id)
-            self.reset_locations(domain, assigned_location_ids)
+                self.set_location(domain, primary_location_id, commit=False)
+            self.reset_locations(domain, assigned_location_ids, commit=False)
```
Comment (Jtang-1) [inline corehq/apps/users/models.py:566]: I looked and it doesn't seem to the case. But just noting that we need to make sure to check there isn't temporal coupling that requires location to be committed here

**Review by Jtang-1 — APPROVED** [review]
no blocking comments, good work investigating this. Looked like a tricky bug

File: corehq/apps/users/views/web.py [hunk 5]
```diff
@@ -181,10 +181,11 @@ def __call__(self, request, uuid, **kwargs):
                         created_via=USER_CHANGE_VIA_INVITATION,
                         domain=invitation.domain,
                         is_domain_admin=False,
+                        commit=False
                     )
-                    user.save()
```
Comment (Jtang-1) [inline corehq/apps/users/views/web.py:185]: I'm not seeing where the user gets saved now if `commit=False` is passed down the call stack?

**Comment by esoergel** [comment]
Just spent some time reviewing the code area before looking over the changes here, since it sounds like there's still uncertainty about the problem.  I have a few thoughts.

> reducing the number of saves to couch 

That's definitely helpful, though I think with a resource conflict, I'd guess the problem is the number of _reads_ from couch.  If you have a single user instance on a single thread and save a bunch of times in sequence, it should be fine.  The resource conflict is essentially a concurrency issue.  Some ideas on how that might come up:

* A user save or something else kicks off an async process (pillows, celery) that updates the user while the main thread is still processing.  Now the main thread has an out-of-date user and it'll get a conflict if it tries to save.  Reducing the number of saves does seem like it should help here.
* A user is created, then two threads are started with the same user instance (pass by value).  If both threads save the user, the second one will have a resource conflict
* Something fetches a cached version of the user doc and saves that.  If that doc is out of date, this will raise a resource conflict.  Things that are going to save the user should use `strict` or `skip_cache` or whatever it's called.
* Multiple instances of the user object are being passed around.  It looks like the main invitation code passes around the same user instance, but I wouldn't be surprised to see some secondary process look up the user object and save it.

Here's an illustration of that last point:
```python
>>> user = CommCareUser.get_by_domain('ethan-test')[0]
>>> user.phone_numbers
[]
>>> user.phone_numbers.append('1')
>>> user.save()
>>> same_user = CommCareUser.get(user._id)
>>> same_user.phone_numbers.append('2')
>>> same_user.save()
>>> same_user.phone_numbers
['1', '2']
>>> user.phone_numbers
['1']
>>> user.phone_numbers.append('3')
>>> user.save()
    ...
    raise ResourceConflict
```

> When an error is triggered in add_as_web_user, it seems to stop execution after the CouchUser was created but before the fields were populated from the invitation object, causing the end-user issue.

It looks like the user is saved before it gets to that function, actually:
https://github.com/dimagi/commcare-hq/blob/1a70089d74f4c7c367799ef83e7eddcd2739340a/corehq/apps/users/views/web.py#L178-L185

One thing that might be helpful here is to throw some print statements in `user.save()` (or a `import traceback; traceback.print_stack()`) and make sure all of those saves are accounted for.  Things happening in celery could certainly make this hard to debug - it'd be good to know whether there's anything there that saves the user.  Another thing that might help is to move as much as possible to background tasks, so the main thread is relatively clean.  Having access to actual DB locking and atomic transactions would make this way easier.

File: corehq/apps/users/tests/test_user_invitation.py [hunk 6]
```diff
@@ -98,3 +116,31 @@ def test_redirect_if_invite_email_does_not_match(self):
...
+        Invitation.objects.create(
+            email='test5@dimagi.com',
+            uuid=invite_uuid,
+            domain=self.domain,
+            is_accepted=False,
+            invited_on=(datetime.datetime.now() - datetime.timedelta(days=1)),
+            invited_by="system@dimagi.com",
+            role=self.role1.get_qualified_id()
+        )
+        response = UserInvitationView()(request, invite_uuid, domain=self.domain)
```
Comment (esoergel) [inline corehq/apps/users/tests/test_user_invitation.py:155]: :star:

File: corehq/apps/reports/util.py (hunk 3 above)
Comment (esoergel) [reply corehq/apps/reports/util.py:515]: Though if the param is never set to `True`, then you could probably just remove it, right?

**Comment by esoergel** [comment]
Know if it'd be doable/advisable to move the tableau stuff to a celery task?  There's a lot going on on user save, and it'd be nice to process some of that in the background.

File: corehq/apps/reports/util.py (hunk 2 above)
Comment (Jtang-1) [reply corehq/apps/reports/util.py:532]: oh right, duh

File: corehq/apps/users/models.py (hunk 4 above)
Comment (Jtang-1) [reply corehq/apps/users/models.py:566]: just looking at this comment for `reset_locations` 
```            
This should be called after updating primary location via set_location/unset_location
If primary-location is not set, then next available location from
assigned_location_ids is set as the primary-location
```
I thought there was a possibility that it requires `location_id` to be saved first. but doesn't seem like it

**Review by esoergel — APPROVED** [review]

**Pull request merged**